mender.go

```diff
@@ -104,20 +132,61 @@ func (m mender) GetCurrentImageID() string {
 	return imageID
 }
 
-func (m *mender) updateState() error {
+func (m *mender) changeState(state MenderState) {
+	log.Infof("Mender state: %v -> %v", m.state, state)
+	m.state = state
+}
+
+func (m *mender) hasUpgrade() (bool, error) {
 	env, err := m.ReadEnv("upgrade_available")
 	if err != nil {
-		return err
+		return false, err
 	}
 	upgradeAvailable := env["upgrade_available"]
 
 	// we are after update
 	if upgradeAvailable == "1" {
-		m.state = MenderRunningWithFreshUpdate
-		return nil
+		return true, nil
 	}
-	m.state = MenderFreshInstall
-	return nil
+	return false, nil
+}
+
+func (m *mender) updateState() {
+
+	newstate := MenderStateUnknown
+	var merr error
+
+	switch m.state {
```

Comment:
I am not sure if there is not some logic error here. Wont we always start with `MenderStateInit` (which is default state in constructor)? Then, it looks like `needsBootstrap()` will be always checked here, and we will end up with `MenderStateBootstrapped`.
If there are other places in the code relaying on the actual state check we will report wrong state as the on we are really in might be `MenderStateRunningWithFreshUpdate` or `MenderStateWaitForUpdate` which will be reported only after entering `updateState()` for the second time.

What is more `needsBootstrap()` might not work well. What if we've been already bootstrapped, then update and reset happened. We will end up running from inactive partition being not bootstrapped.